tests/test_traits_core.cc

```diff
@@ -0,0 +1,317 @@
+protected:
+  Fit<HasProtectedValidFitImpl, X> fit(const std::vector<X> &,
+                                          const MarginalDistribution &) const {
+    return {};
+  };
+};
+
+class HasPrivateValidFitImpl : public ModelBase<HasPrivateValidFitImpl>  {
+private:
+  Fit<HasPrivateValidFitImpl, X> fit(const std::vector<X> &,
+                                        const MarginalDistribution &) const {
+    return {};
+  };
+};
+
+class HasValidAndInvalidFitImpl : public ModelBase<HasValidAndInvalidFitImpl> {
+public:
+  Fit<HasValidAndInvalidFitImpl, X> fit(const std::vector<X> &,
+                                        const MarginalDistribution &) const {
+    return {};
+  };
+
+  Fit<HasValidAndInvalidFitImpl, X> fit(const std::vector<X> &,
+                                        const MarginalDistribution &) {
+    return {};
+  };
+};
+
+class HasValidXYFitImpl : public ModelBase<HasValidXYFitImpl> {
+public:
+  Fit<HasValidXYFitImpl, X> fit(const std::vector<X> &,
+                                const MarginalDistribution &) const {
+    return {};
+  };
+
+  Fit<HasValidXYFitImpl, Y> fit(const std::vector<Y> &,
+                                      const MarginalDistribution &) const {
+    return {};
+  };
+};
+
+class HasNoFitImpl {};
+
+TEST(test_traits_core, test_has_valid_fit) {
+  EXPECT_TRUE(bool(has_valid_fit<HasValidFitImpl, X>::value));
+  EXPECT_FALSE(bool(has_valid_fit<HasWrongReturnTypeFitImpl, X>::value));
+  EXPECT_FALSE(bool(has_valid_fit<HasNonConstFitImpl, X>::value));
+  EXPECT_FALSE(bool(has_valid_fit<HasNonConstArgsFitImpl, X>::value));
+  EXPECT_FALSE(bool(has_valid_fit<HasProtectedValidFitImpl, X>::value));
+  EXPECT_FALSE(bool(has_valid_fit<HasPrivateValidFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_valid_fit<HasValidAndInvalidFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_valid_fit<HasValidXYFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_valid_fit<HasValidXYFitImpl, Y>::value));
+  EXPECT_FALSE(bool(has_valid_fit<HasNoFitImpl, X>::value));
+  //  EXPECT_TRUE(bool(has_valid_fit<HasInheritedFitImpl, Y>::value));
+}
+
+TEST(test_traits_core, test_has_possible_fit_impl) {
+  EXPECT_TRUE(bool(has_possible_fit<HasValidFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_possible_fit<HasWrongReturnTypeFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_possible_fit<HasNonConstFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_possible_fit<HasNonConstArgsFitImpl, X>::value));
+  EXPECT_FALSE(bool(has_possible_fit<HasProtectedValidFitImpl, X>::value));
+  EXPECT_FALSE(bool(has_possible_fit<HasPrivateValidFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_possible_fit<HasValidAndInvalidFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_possible_fit<HasValidXYFitImpl, X>::value));
+  EXPECT_TRUE(bool(has_possible_fit<HasValidXYFitImpl, Y>::value));
+  EXPECT_FALSE(bool(has_possible_fit<HasNoFitImpl, X>::value));
+}
+
+TEST(test_traits_core, test_fit_type) {
+  EXPECT_TRUE(bool(std::is_same<Z, fit_type<FitModel<X, Z>, Y>::type>::value));
```

Comment:
Some failure cases would be good here.

Reply:
Currently calling `fit_type<Model, Feature>` for a model which doesn't have a valid fit for that feature results in a compile error which, I think, is desirable but makes testing failure modes difficult.

Speaking of which, are you aware of any ways of unit testing for situations that you want to make sure don't compile?